File: 0877-stone-game/0877-stone-game.py
class Solution:
    def stoneGame(self, piles: List[int]) -> bool:
        n = len(piles)
        dp = piles[:]
        for i in range(1, n):
            for j in range(0, n - i):
                dp[j] = max(piles[j] - dp[j+1], piles[j+i] - dp[j])
        return dp[0] > 0


# class Solution:
#     def stoneGame(self, piles: List[int]) -> bool:
#         return True

File: LeetCode/0877-stone-game/0877-stone-game.java
class Solution {
    public boolean stoneGame(int[] piles) {
        int n = piles.length;
        int[] dp = new int[n];
        for (int i = 0; i < n; i++) {
            dp[i] = piles[i];
        }
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < n - i; j++) {
                dp[j] = Math.max(piles[j] - dp[j+1], piles[j+i] - dp[j]);
            }
        }
        return dp[0] > 0;
    }
}




// class Solution {
//     public boolean stoneGame(int[] piles) {
//         return true;
//     }
// }